source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -336,10 +336,39 @@ namespace
         return elem;
     }
 
-    void readDocument(const DocumentPtr& doc, PrvStage* stage, RtFileIo::ReadFilter filter)
+    void readSourceUri(const DocumentPtr& doc, PrvStage* stage)
     {
+        StringSet uris = doc->getReferencedSourceUris();
+        RtTokenList& stageUris = stage->getSourceUri();
+        for (const string& uri : uris)
+        {
+            stageUris.push_back(RtToken(uri));
+        }
+    }
+
+    void readDocument(const DocumentPtr& doc, PrvStage* stage, const RtReadOptions* readOptions)
+    {
+        // Set the source location 
+        const std::string& uri = doc->getSourceUri();
+        RtTokenList uris = stage->getSourceUri();
+        uris.push_back(RtToken(uri));
```

Comment:
I think `&` for reference is missing here, so `uris` is a copy of the list and push_back till only push to the local copy.

Reply:
An option would be to let `getSourceUri` be a const method, and add an additional `addSourceUri` for adding new file URI paths.